### PR TITLE
Skip button

### PR DESCRIPTION
**Pull request by Uppun**
time for refactoring

### CODE REVIEW

File: src/radio/index.js [hunk 1]
```diff
@@ -150,7 +150,7 @@ class Radio extends EventEmitter {
     const total = this.order.length;
     const needed = Math.ceil(ratio * total);
     this.emit('skips', this.skips, needed);
-    if (needed != -1 && this.skips.size >= needed) {
+    if (needed !== -1 && this.skips.size >= needed) {
```
Comment (meishuu) [inline src/radio/index.js]: can `needed` ever be -1?

File: src/radio/index.js [hunk 2]
```diff
@@ -138,19 +138,22 @@ class Radio extends EventEmitter {
...
-
-    this.skips.add(user.id);
     this.checkSkips();
+    this.skips.add(user.id);
     return true;
   }
 
   checkSkips() {
-    if (!this.current) return false;
+    if (!this.current) return -1;
```
Comment (meishuu) [inline src/radio/index.js]: why this change?

File: src/radio/index.js [hunk 3]
```diff
@@ -138,19 +138,22 @@ class Radio extends EventEmitter {
     }
 
     if (!this.order.some(u => u.equals(user))) return false;
-
-    this.skips.add(user.id);
     this.checkSkips();
+    this.skips.add(user.id);
```
Comment (meishuu) [inline src/radio/index.js]: are we sure we only want to add the vote *after* checking?

File: src/discord/index.js [hunk 4]
```diff
@@ -134,6 +134,15 @@ export default class Bot extends EventEmitter {
             .catch(reason => message.reply(`Oops! Something went wrong! Error: ${reason}`));
           break;
         }
+        case '!skip': {
```
Comment (meishuu) [inline src/discord/index.js]: blank line before and after case block

File: src/discord/index.js [hunk 5]
```diff
@@ -134,6 +134,15 @@ export default class Bot extends EventEmitter {
...
+          const res = radio.voteSkip(message.author);
+          if (!res) {
+            message.reply(`Error: You must be in the voice channel to vote to skip songs!`);
+            break;
```
Comment (meishuu) [inline src/discord/index.js]: can just
```js
if (...) {
  message.reply(...);
} else {
  message.reply(...);
}
break;
```

File: src/discord/index.js [hunk 6]
```diff
@@ -164,6 +173,8 @@ export default class Bot extends EventEmitter {
       if (song && this.voiceConnection) {
         const dispatcher = this.voiceConnection.playFile(fp, {volume: 0});
         dispatcher.setVolumeDecibels(BASELINE_DB + song.gain);
+      } else if (this.voiceConnection && this.voiceConnection.dispatcher) {
```
Comment (meishuu) [inline src/discord/index.js]: so, the branch above this one also relies on `this.voiceConnection` being nonnull

you can probs think of how this can be refactored to simplify this logic

File: src/radio/index.js [hunk 7]
```diff
@@ -138,19 +138,22 @@ class Radio extends EventEmitter {
...
 
   checkSkips() {
-    if (!this.current) return false;
+    if (!this.current) return -1;
 
     const ratio = skipRatio(this.current.duration);
     const total = this.order.length;
     const needed = Math.ceil(ratio * total);
     this.emit('skips', this.skips, needed);
+    if (needed !== -1 && this.skips.size >= needed) {
```
Comment (meishuu) [inline src/radio/index.js]: i dont think `needed` can ever be -1

File: src/web/app.js [hunk 8]
```diff
@@ -18,6 +18,10 @@ export default function webapp(web: Web) {
     io.emit('history', history);
   });
 
+  radio.on('skips', (skips, needed) => {
+    io.emit('skips', Array.from(skips), needed);
```
Comment (meishuu) [inline src/web/app.js]: we don’t need to send `needed` anymore

File: src/web/app.js [hunk 9]
```diff
@@ -100,6 +104,10 @@ export default function webapp(web: Web) {
       radio.removeSong(user, qid);
     });
 
+    socket.on('voteSkip', () => {
```
Comment (meishuu) [inline src/web/app.js]: nit: i prefer just calling this `”skip”`

File: web/src/assets/css/todo.css [hunk 10]
```diff
@@ -310,8 +310,11 @@ h2 {
   padding: 0 8px;
 }
 
+.inactive {
```
Comment (meishuu) [inline web/src/assets/css/todo.css]: this should be specifically on `.skip span` and should be placed below the standalone `.skip span` style

File: web/src/components/CurrentSong.js [hunk 11]
```diff
@@ -2,11 +2,17 @@ import React from 'react';
...
 
 import {UserAvatar} from './DiscordIcon';
 import SongProgress from './SongProgress';
 
-function CurrentSong({song, dj, startTime, offset}) {
+function CurrentSong({song, dj, startTime, offset, skips, doVoteSkip, members}) {
+  const minutes = song.duration / 60;
+  const skipThreshold = 0.6 - 0.3 / (1 + Math.exp(3 - minutes / 3));
+  const votesNeeded = Math.ceil(skipThreshold * members.length);
+  const skipTextClassName = classNames({inactive: !skips});
```
Comment (meishuu) [inline web/src/components/CurrentSong.js]: newline after

plis whitespace more

File: web/src/components/CurrentSong.js [hunk 12]
```diff
@@ -47,15 +53,31 @@ function CurrentSong({song, dj, startTime, offset}) {
           <span className="dj-name">{dj.name}</span>
         </div>
         <div className="skip">
-          <button className="btn" disabled>
+          <button className="btn" onClick={doVoteSkip}>
             Skip
           </button>
-          <span>{/* TODO */}</span>
+          <span className={skipTextClassName}>
+            {skips && skips > 0 ? `${skips}/${votesNeeded} votes` : `0/${votesNeeded} votes`}
```
Comment (meishuu) [inline web/src/components/CurrentSong.js]: `skips > 0` implies `skips`

Comment (meishuu) [reply web/src/components/CurrentSong.js]: also you should make it just default 0

File: web/src/components/CurrentSong.js [hunk 13]
```diff
@@ -47,15 +53,31 @@ function CurrentSong({song, dj, startTime, offset}) {
...
+  };
+}
+
+function mapDispatchToProps(dispatch) {
+  return {
+    doVoteSkip() {
+      dispatch(voteSkip());
+    },
+  };
+}
```
Comment (meishuu) [inline web/src/components/CurrentSong.js:92]: newline after

File: src/radio/index.js (hunk 2 above)
Comment (Uppun) [reply src/radio/index.js]: Oh this was done because I was returning a number for voteskip to check on skip because, at the time, the skip was being done there. So I needed to have it return a number and it would use that to check.

**Review by meishuu — COMMENTED** [review]
high level considerations:

- when you vote skip, the web button stys blue and clickable, text channel says it registers vote. either we should let you revoke a vote, or disable the button and say you already voteskipped

- doesn’t look like topic supports this yet? is this for now or a future pr? i guess it doesn’t help topic is currently broken atm

File: src/radio/index.js (hunk 2 above)
Comment (Uppun) [reply src/radio/index.js]: Also I suppose I can just go back now

File: src/radio/index.js (hunk 7 above)
Comment (Uppun) [reply src/radio/index.js]: Yeah the above comment covered this

**Comment by Uppun** [comment]
I think we should handle the topic thing when we actually work on fixing topic. So a future PR would probably be best.

Also vote-revoking sounds like a decent change. I suppose we could have it check and see if you're already in the skips array and if so it could just do a set.delete

File: src/radio/index.js [hunk 14]
```diff
@@ -138,9 +138,9 @@ class Radio extends EventEmitter {
     }
 
     if (!this.order.some(u => u.equals(user))) return false;
-
     this.skips.add(user.id);
     this.checkSkips();
+
```
Comment (meishuu) [inline src/radio/index.js]: <sub>super turbo ultra mega hyper nit: can we go back to the old spacing we had earlier? moving this newline above the `skips.add`. i like that one better</sup>

File: web/src/components/CurrentSong.js [hunk 15]
```diff
@@ -47,15 +54,32 @@ function CurrentSong({song, dj, startTime, offset}) {
           <span className="dj-name">{dj.name}</span>
         </div>
         <div className="skip">
-          <button className="btn" disabled>
+          <button className="btn" onClick={doVoteSkip}>
             Skip
           </button>
-          <span>{/* TODO */}</span>
+          <span className={skipTextClassName}>
+            {skips > 0 ? `${skips}/${votesNeeded} votes` : `0/${votesNeeded} votes`}
```
Comment (meishuu) [inline web/src/components/CurrentSong.js]: bumping: `skips` should default 0

File: src/radio/index.js [hunk 16]
```diff
@@ -151,6 +151,10 @@ class Radio extends EventEmitter {
     const total = this.order.length;
     const needed = Math.ceil(ratio * total);
     this.emit('skips', this.skips, needed);
+    if (this.skips.size >= needed) {
+      this.getNext();
+    }
+    return true;
```
Comment (meishuu) [inline src/radio/index.js]: is this return value being used anywhere?

**Review by meishuu — COMMENTED** [review]
if we want to support vote revoking then my only concern is that we have proper messaging for whether a user has already voted

on discord side we can kinda do whatevs and just have `!skip` respond with smth saying either added or removed vote (or error case)

on web side we might consider something like, button says `Skip` and vote count stays gray *as long as the user hasn't voted to skip*, but then maybe we switch it to `Unskip`(???) and make the count bright if the user has a skip vote already

File: web/src/components/CurrentSong.js (hunk 15 above)
Comment (meishuu) [reply web/src/components/CurrentSong.js]: this means:
1. you don't have to special case this line for 0 skips
1. you will need to update the reducer

**Comment by meishuu** [comment]
also, are skip votes transmitted to web client on load? i didn't see it in the code and i don't remember from testing

File: src/radio/index.js (hunk 14 above)
Comment (Uppun) [reply src/radio/index.js]: 🐇

File: web/src/components/CurrentSong.js (hunk 15 above)
Comment (Uppun) [reply web/src/components/CurrentSong.js]: 🐇

File: src/radio/index.js (hunk 16 above)
Comment (Uppun) [reply src/radio/index.js]: 🐇

File: web/src/assets/css/todo.css [hunk 17]
```diff
@@ -310,11 +310,15 @@ h2 {
   padding: 0 8px;
 }
 
+
```
Comment (meishuu) [inline web/src/assets/css/todo.css]: extra line?

File: web/src/components/CurrentSong.js [hunk 18]
```diff
@@ -2,11 +2,13 @@ import React from 'react';
...
 import {connect} from 'react-redux';
 import {getCurrentSong} from '../reducers/songs/current';
+import {getMemberList} from '../reducers/members';
+import {voteSkip} from '../actions';
 
 import {UserAvatar} from './DiscordIcon';
 import SongProgress from './SongProgress';
 
-function CurrentSong({song, dj, startTime, offset}) {
+function CurrentSong({song, dj, startTime, offset, skips = 0, doVoteSkip, members}) {
```
Comment (meishuu) [inline web/src/components/CurrentSong.js]: can remove this `= 0` now

File: web/src/reducers/songs/current.js [hunk 19]
```diff
@@ -1,25 +1,32 @@
...
     offset: song.player.startTime - song.player.currentTime,
+    skips: state.skips || 0,
   };
 }
 
 export default function reduceCurrent(state = null, action, songs) {
   switch (action.type) {
     case INITIALIZE:
-      return reduceSong(action.payload.current);
+      return reduceSong(action.payload.current, state);
```
Comment (meishuu) [inline web/src/reducers/songs/current.js]: is `INITIALIZE` sending skips atm?

Comment (meishuu) [reply web/src/reducers/songs/current.js]: actually, you wouldn't be using it with this current code even if it did

File: web/src/reducers/songs/current.js [hunk 20]
```diff
@@ -1,25 +1,32 @@
...
 
     case SET_CURRENT_SONG:
-      return reduceSong(action.payload);
+      return reduceSong(action.payload, state);
```
Comment (meishuu) [inline web/src/reducers/songs/current.js]: since `reduceSong` retains `skips` from state, this means new songs will retain the number of skips as the previous song

File: web/src/reducers/songs/current.js [hunk 21]
```diff
@@ -1,25 +1,32 @@
...
 
-function reduceSong(song) {
+function reduceSong(song, state) {
   if (!song) return null;
   return {
     song: getUid(song),
     dj: song.player.dj.id,
     startTime: song.player.startTime,
     offset: song.player.startTime - song.player.currentTime,
+    skips: state.skips || 0,
```
Comment (meishuu) [inline web/src/reducers/songs/current.js]: also this should be an array 🤔 
EDIT: no it shouldn't

File: web/src/reducers/songs/current.js [hunk 22]
```diff
@@ -1,25 +1,34 @@
...
 export default function reduceCurrent(state = null, action, songs) {
   switch (action.type) {
     case INITIALIZE:
-      return reduceSong(action.payload.current);
+      return reduceSong(action.payload.current, action.payload.skips);
 
     case SET_CURRENT_SONG:
-      return reduceSong(action.payload);
+      return {
+        ...reduceSong(action.payload),
```
Comment (meishuu) [inline web/src/reducers/songs/current.js]: don’t need to spread since we don’t need a copy and aren’t adding properties

File: web/src/reducers/songs/current.js [hunk 23]
```diff
@@ -1,25 +1,34 @@
...
 
-function reduceSong(song) {
+function reduceSong(song, radioSkips) {
   if (!song) return null;
   return {
     song: getUid(song),
     dj: song.player.dj.id,
     startTime: song.player.startTime,
     offset: song.player.startTime - song.player.currentTime,
+    skips: radioSkips || 0,
```
Comment (meishuu) [inline web/src/reducers/songs/current.js]: two things:

1. since we’re using a function arg for this, we can just default it to 0

2. if we rename it `skips`, this line can just be `skips,`

**Review by meishuu — COMMENTED** [review]
this is better :blobcatmelt:

File: web/src/reducers/songs/current.js (hunk 23 above)
Comment (Uppun) [reply web/src/reducers/songs/current.js]: i need to get better at visual code efficiency

Comment (Uppun) [reply web/src/reducers/songs/current.js]: also :rabbit2:

File: web/src/reducers/songs/current.js (hunk 22 above)
Comment (Uppun) [reply web/src/reducers/songs/current.js]: :rabbit2:

File: web/src/components/CurrentSong.js [hunk 24]
```diff
@@ -47,15 +54,30 @@ function CurrentSong({song, dj, startTime, offset}) {
           <span className="dj-name">{dj.name}</span>
         </div>
         <div className="skip">
-          <button className="btn" disabled>
+          <button className="btn" onClick={doVoteSkip}>
             Skip
           </button>
-          <span>{/* TODO */}</span>
+          <span className={skipTextClassName}>{`${skips}/${votesNeeded} votes`}</span>
```
Comment (meishuu) [inline web/src/components/CurrentSong.js]: `<span (...)>{skips}/{votesNeeded} votes</span>`

**Comment by Uppun** [comment]
Okay so I'm not sure what the problem is but the
`<span (...)>{skips}/{votesNeeded} votes</span>` 
line made travis throw a fit. I undid it and re-pushed and now travis likes it.

File: src/radio/index.js [hunk 25]
```diff
@@ -138,6 +138,11 @@ class Radio extends EventEmitter {
     }
 
     if (!this.order.some(u => u.equals(user))) return false;
+    if (this.skips.has(user.id)) {
```
Comment (meishuu) [inline src/radio/index.js]: nit: newline above

File: src/radio/index.js [hunk 26]
```diff
@@ -138,6 +138,11 @@ class Radio extends EventEmitter {
...
+      this.skips.delete(user.id);
+      this.emit('unskip', this.skips);
```
Comment (meishuu) [inline src/radio/index.js]: this event looks identical to `skips`

File: src/radio/index.js [hunk 27]
```diff
@@ -150,7 +155,10 @@ class Radio extends EventEmitter {
     const ratio = skipRatio(this.current.duration);
     const total = this.order.length;
     const needed = Math.ceil(ratio * total);
-    this.emit('skips', this.skips, needed);
+    this.emit('skips', this.skips);
```
Comment (meishuu) [inline src/radio/index.js]: i think this should be part of `voteSkip()` instead

the only question is making sure we're correctly handling cases that used to directly call `checkSkips()`

File: web/src/components/CurrentSong.js [hunk 28]
```diff
@@ -2,11 +2,13 @@ import React from 'react';
...
 import {connect} from 'react-redux';
 import {getCurrentSong} from '../reducers/songs/current';
+import {getMemberList} from '../reducers/members';
+import {voteSkip} from '../actions';
 
 import {UserAvatar} from './DiscordIcon';
 import SongProgress from './SongProgress';
 
-function CurrentSong({song, dj, startTime, offset}) {
+function CurrentSong({song, dj, startTime, offset, skips, doVoteSkip, members}) {
```
Comment (meishuu) [inline web/src/components/CurrentSong.js]: nit: `doVoteSkip` after `members` so we have state/props first and actions last

File: web/src/components/CurrentSong.js [hunk 29]
```diff
@@ -47,15 +54,32 @@ function CurrentSong({song, dj, startTime, offset}) {
           <span className="dj-name">{dj.name}</span>
         </div>
         <div className="skip">
-          <button className="btn" disabled>
+          <button className="btn" onClick={doVoteSkip}>
             Skip
```
Comment (meishuu) [inline web/src/components/CurrentSong.js]: wdyt about the idea of making this button say "Unskip" if the user has voteskipped, and also keeping the vote count gray ("inactive") and only making it active if the user has voteskipped?

**Review by meishuu — COMMENTED** [review]
nits and some UX questions

i wonder what sort of reply we should make for the discord `!skip` command but also idc too much atm about handling discord side

**Comment by Uppun** [comment]
Oh I never replied to your discord nit/ux question.

Well, I suppose one thing we could do if we want a discord chat message is we could come up with a new emit that will just tell the discord side to either say "Skip vote added!" and "Skip vote removed!" based on which it was. Could put that in voteSkip in the if/else block.

Outside of that I was thinking part of the feedback would of course be on the setTopic thing when we get around to fixing that.

File: src/radio/index.js [hunk 30]
```diff
@@ -139,8 +139,14 @@ class Radio extends EventEmitter {
...
 
-    this.skips.add(user.id);
+    if (this.skips.has(user.id)) {
+      this.skips.delete(user.id);
+    } else {
+      this.skips.add(user.id);
+    }
+
     this.checkSkips();
+    this.emit('skips', this.skips);
```
Comment (meishuu) [inline src/radio/index.js]: checkSkips may advance the song; does this have the potential to emit a useless skips event after a song gets skipped?

File: src/discord/index.js [hunk 31]
```diff
@@ -134,6 +134,15 @@ export default class Bot extends EventEmitter {
             .catch(reason => message.reply(`Oops! Something went wrong! Error: ${reason}`));
           break;
         }
+
+        case '!skip': {
+          const res = radio.voteSkip(message.author);
+          if (!res) {
```
Comment (meishuu) [inline src/discord/index.js]: tbqh i think i might’ve said otherwise but we should at least reply with *something* just so the user gets immediate confirmation that the command worked

**Review by meishuu — COMMENTED** [review]
almost there :blobcatmelt:

will web client properly update vote count when a dj leaves?

File: src/radio/index.js (hunk 30 above)
Comment (Uppun) [reply src/radio/index.js]: Good point. I suppose I should put it prior to the checkSkips call, especially considering when we had the emit in check skips it was doing it prior to moving to the next song as well.

**Comment by Uppun** [comment]
The only thing removeDJ seemed to be using checkSkips for was to update the skip count. iirc it did so properly prior to removing the checkSkips call but just slapping the one line it actually cared about in there so it could emit on it's own should ultimately do the same thing. The rest of checkSkips was for actually skipping to the next song. The removal of a DJ (that isn't the dj of the current song) shouldn't ever trigger the next song movement so most of the function is unnecessary.

File: src/discord/index.js (hunk 31 above)
Comment (Uppun) [reply src/discord/index.js]: I'll throw in a somewhat generic message for now. I do agree that having a reply is important because I noticed Virani was confused as to whether or not it worked when he did it without the web application up. 

We'll also be doing the setTopic thing but that kind of update is not right in your face confirmation since it's kinda tucked away at the top in small text with a bunch of other stuff. I'd personally like to be able to differentiate between skips and unskips on the message as well. How would you feel about me changing the return value? We could do something like -1, 0, 1 where -1 is the response to trying to skip when not a DJ, 0 is skip and 1 is unskip.

**Comment by meishuu** [comment]
i see; i couldn’t expand to see the full context of that line change on phone

>The removal of a DJ (that isn't the dj of the current song) shouldn't ever trigger the next song movement

why not? if we hit the vote threshold from someone leaving, enforcing this means someone has to unvote then revote to get the skip to proc

File: src/discord/index.js (hunk 31 above)
Comment (meishuu) [reply src/discord/index.js]: another thing you could do is return an object

you could do some `{voteSkipped: false}` for instance, which as a whole is truthy so it passes error check but then you could inspect the property

**Comment by Uppun** [comment]
Oh I see, I didn't think of it from the perspective of that. Well then I could easily just do an emit -> checkSkips call.

File: src/discord/index.js (hunk 31 above)
Comment (Uppun) [reply src/discord/index.js]: Oh, and otherwise I could just pass nothing for an error?

File: src/discord/index.js [hunk 32]
```diff
@@ -134,6 +134,21 @@ export default class Bot extends EventEmitter {
...
+            message.reply(`Error: You must be in the voice channel to vote to skip songs!`);
+            break;
+          }
+          if (!res.voteSkipped) {
+            message.reply(`Vote skip removed!`);
+            break;
```
Comment (meishuu) [inline src/discord/index.js]: nit: i prefer
```js
if (res.voteSkipped) {
  // added
} else {
  // removed
}
break;
```

File: src/web/app.js [hunk 33]
```diff
@@ -116,6 +124,7 @@ export default function webapp(web: Web) {
         queue: queue.map(({fp, ...item}) => item), // eslint-disable-line no-unused-vars
         current,
         history: radio.history,
+        skips: radio.skips.size,
```
Comment (meishuu) [inline src/web/app.js]: is this correct? i thought we were sending down the array now

**Review by meishuu — COMMENTED** [review]
soon:tm:

File: src/discord/index.js [hunk 34]
```diff
@@ -134,6 +134,21 @@ export default class Bot extends EventEmitter {
...
           break;
         }
+
+        case '!skip': {
+          const res = radio.voteSkip(message.author);
+          if (!res) {
+            message.reply(`Error: You must be in the voice channel to vote to skip songs!`);
+            break;
+          }
+          if (res.voteSkipped) {
```
Comment (meishuu) [inline src/discord/index.js:145]: nit: newline above

File: web/src/assets/css/todo.css [hunk 35]
```diff
@@ -311,10 +311,13 @@ h2 {
 }
 
 .skip span {
-  display: none;
   margin-left: 6px;
 }
 
+.skip span.inactive {
+  opacity: 0.4;
```
Comment (meishuu) [inline web/src/assets/css/todo.css]: i still want to try 0.3 :blobcatverysad:

File: web/src/socket.js [hunk 36]
```diff
@@ -58,4 +61,8 @@ export default store => {
   socket.on('add status', queueItem => {
     store.dispatch(actions.setSongStatus(queueItem));
   });
+
+  socket.on('skips', skips => {
+    store.dispatch(actions.setSkipStatus({skips: skips}));
```
Comment (meishuu) [inline web/src/socket.js]: `{skips}`

**Review by meishuu — APPROVED** [review]
just nits, otherwise lgtm!
